proposals/provenance-logging.md

```diff
@@ -0,0 +1,128 @@
+## Motivation
+Reproducible ML is a [big problem](https://petewarden.com/2018/03/19/the-machine-learning-reproducibility-crisis/). 
+Kubeflow could add a lot of values to users by making it easy for users to track how datasets and models were produced.
+A lot of valuable information is available in K8s resource specifications. Resources like K8s Jobs, TFJobs, ConfigMaps, and Argo workflows
+often contain information (e.g. via command line arguments) about inputs and outputs and the code (docker image) used. There are
+two major problems to using this information to track provenance of models and data today:
+
+1. The information is not durable
+
+	* I should not rely on the K8s APIServer as a persistent, historical data store for resources.
+
+1. The schema is not convenient
+
+	* People use a variety of conventions/patterns to pass inputs/outputs to their programs
+	* Lack of uniformity is a barrier to building tools that can surface information in useful ways.
+
+The purpose of this proposal is to solve these problems to facilitate tools to track and surface provenance.
+
+## Goals
+
+1. Build an agent to persist resource specs to durable storage
+
+1. Allow user defined ETL pipelines to transform raw data into useful tables.
+
+## Non-Goals
+
+1. Building a comprehensive set of tools to surface provenance information.
+
+
+## Design
+
+As illustrated by the diagram below the system will consist of
+
+1. A cluster agent that watches the APIServer and records K8s objects to a persistent datastore
+1. A row database (e.g. HBASE/BigTable) to store the raw K8s objects
+1. One or more ETL pipelines to turn the raw data into transaction records
+1. A database to store the transaction records
```

Comment:
I wonder if we can add hooks to let other things plug in here. E.g. could this be the start of the pipeline auditing?

Reply:
Isn't it already inherently extensible? Anyone can an ETL job to produce transaction records or consumer the transaction records.